### PR TITLE
feat(pubsub): an endurance benchmark

### LINKED ISSUE #4564: Implement an "endurance" test
Write a test that exercises the library and runs for hours. This is to trigger rare leaks, race conditions, or other problems that go undetected by short tests.

The test may need a separate design doc, but it should:

- Run with multiple threads
- Randomly publish messages
- Receive messages and randomly ack or nack them

As part of the CI build we would wish to run a very short version of this test, just to ensure the code compiles and does not crash in an embarrassing way. Running the test for 24 hours or so would be a manual step.

### PR DESCRIPTION
**Pull request by coryan**
Verify the Pub/Sub client library works well when used for hours.

Fixes #4564

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5225)

### CODE REVIEW

File: google/cloud/pubsub/benchmarks/endurance.cc [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+        flow_control.Debug(std::cout);
+        std::cout << std::endl;
+      } else {
+        auto const p = std::minmax_element(samples.begin(), samples.end());
+        auto const sum = std::accumulate(samples.begin(), samples.end(),
+                                         std::chrono::microseconds{0});
+        auto const mean = [](std::chrono::microseconds sum, std::size_t n) {
+          sum /= n;
+          return sum;
+        }(sum, samples.size());
```
Comment (devbww) [inline google/cloud/pubsub/benchmarks/endurance.cc:315]: Why not simply `auto const mean = sum / samples.size()`?

File: google/cloud/pubsub/benchmarks/endurance.cc [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+  pubsub::TopicAdminClient topic_admin(pubsub::MakeTopicAdminConnection());
+  pubsub::SubscriptionAdminClient subscription_admin(
+      pubsub::MakeSubscriptionAdminConnection());
+
+  auto generator = google::cloud::internal::MakeDefaultPRNG();
+
+  auto const configured_topic = config->topic_id;
+
+  // If there is no pre-defined topic for this test, create one and
+  // automatically remove it at the end of the test.
```
Comment (devbww) [inline google/cloud/pubsub/benchmarks/endurance.cc:179]: It looks like we delete the topic unconditionally.  Should be only do that if we created one?

File: google/cloud/pubsub/benchmarks/endurance.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/benchmarks/endurance.cc]: LOL, yes of course.  It used to be more complicated and I failed to simplify it fully.

File: google/cloud/pubsub/benchmarks/endurance.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/benchmarks/endurance.cc:179]: Good catch, fixed.

**Review by devbww — APPROVED** [review]

**Pull request merged**